pkg/blockio/blockio.go

```diff
@@ -0,0 +1,420 @@
+	var errors *multierror.Error
+	oci := cgroups.NewOciBlockIOParameters()
+	for _, dp := range dps {
+		var err error
+		var weight, throttleReadBps, throttleWriteBps, throttleReadIOPS, throttleWriteIOPS int64
+		weight, err = parseAndValidateInt64("Weight", dp.Weight, -1, 10, 1000)
+		errors = multierror.Append(errors, err)
+		throttleReadBps, err = parseAndValidateInt64("ThrottleReadBps", dp.ThrottleReadBps, -1, 0, -1)
+		errors = multierror.Append(errors, err)
+		throttleWriteBps, err = parseAndValidateInt64("ThrottleWriteBps", dp.ThrottleWriteBps, -1, 0, -1)
+		errors = multierror.Append(errors, err)
+		throttleReadIOPS, err = parseAndValidateInt64("ThrottleReadIOPS", dp.ThrottleReadIOPS, -1, 0, -1)
+		errors = multierror.Append(errors, err)
+		throttleWriteIOPS, err = parseAndValidateInt64("ThrottleWriteIOPS", dp.ThrottleWriteIOPS, -1, 0, -1)
+		errors = multierror.Append(errors, err)
+		if dp.Devices == nil {
+			if weight > -1 {
+				oci.Weight = weight
+			}
+			if throttleReadBps > -1 || throttleWriteBps > -1 || throttleReadIOPS > -1 || throttleWriteIOPS > -1 {
+				errors = multierror.Append(errors, fmt.Errorf("ignoring throttling (rbps=%#v wbps=%#v riops=%#v wiops=%#v): Devices not listed",
+					dp.ThrottleReadBps, dp.ThrottleWriteBps, dp.ThrottleReadIOPS, dp.ThrottleWriteIOPS))
+			}
+		} else {
+			blockDevices, err := currentPlatform.configurableBlockDevices(dp.Devices)
+			if err != nil {
+				// Problems in matching block device wildcards and resolving symlinks
+				// are worth reporting, but must not block configuring blkio where possible.
+				log.Error(err.Error())
+			}
+			if len(blockDevices) == 0 {
+				log.Warn("no matches on any of Devices: %v, parameters ignored", dp.Devices)
+			}
+			for _, blockDeviceInfo := range blockDevices {
+				if weight != -1 {
+					if ios, found := currentIOSchedulers[blockDeviceInfo.DevNode]; found {
+						if ios != "bfq" && ios != "cfq" {
+							log.Warn("weight has no effect on device %#v due to "+
+								"incompatible io-scheduler %#v (bfq or cfq required)", blockDeviceInfo.DevNode, ios)
+						}
+					}
+					oci.WeightDevice.Update(blockDeviceInfo.Major, blockDeviceInfo.Minor, weight)
+				}
+				if throttleReadBps != -1 {
+					oci.ThrottleReadBpsDevice.Update(blockDeviceInfo.Major, blockDeviceInfo.Minor, throttleReadBps)
+				}
+				if throttleWriteBps != -1 {
+					oci.ThrottleWriteBpsDevice.Update(blockDeviceInfo.Major, blockDeviceInfo.Minor, throttleWriteBps)
+				}
+				if throttleReadIOPS != -1 {
+					oci.ThrottleReadIOPSDevice.Update(blockDeviceInfo.Major, blockDeviceInfo.Minor, throttleReadIOPS)
+				}
+				if throttleWriteIOPS != -1 {
+					oci.ThrottleWriteIOPSDevice.Update(blockDeviceInfo.Major, blockDeviceInfo.Minor, throttleWriteIOPS)
+				}
+			}
+		}
+	}
+	return oci, errors.ErrorOrNil()
+}
+
+// parseAndValidateInt64 parses quantities, like "64 M", and validates that they are in given range.
```

Comment:
nit: I got confused by the name as it's handling positive integers. Rename to `parseAndValidateQuantity` or smth perhaps?

Reply:
Fixed.